File: src/spatial-hash.js
/**
 * Spatial Hash Grid for fast neighbor queries.
 *
 * Uses the CellStart + ParticleNext linked-list scheme:
 *   - cellStart[cellIndex] = head of the particle chain in that cell
 *   - particleNext[particleIndex] = next particle in the same cell
 *
 * queryState(x, y, radius, state, cb) checks the surrounding cells.
 */
export class SpatialHash {
  constructor(cellSize, worldW, worldH) {
    this.cellSize = cellSize;
    this.cols = Math.ceil(worldW / cellSize);
    this.rows = Math.ceil(worldH / cellSize);
    this.numCells = this.cols * this.rows;

    this.cellStart = new Int32Array(this.numCells);
    this.particleNext = null;
  }

  /**
   * Rebuild the spatial index from current particle positions.
   * Must be called once per tick, before any queries.
   */
  build(state) {
    const n = state.count;

    if (!this.particleNext || this.particleNext.length < n) {
      this.particleNext = new Int32Array(Math.max(n, 1024));
    }

    this.cellStart.fill(-1);

    const cs = this.cellSize;
    const cols = this.cols;

    for (let i = 0; i < n; i++) {
      const col = (state.x[i] / cs) | 0;
      const row = (state.y[i] / cs) | 0;
      const ci = row * cols + col;

      if (ci < 0 || ci >= this.numCells) {
        this.particleNext[i] = -1;
        continue;
      }

      this.particleNext[i] = this.cellStart[ci];
      this.cellStart[ci] = i;
    }
  }

  /**
   * Find all particles within `radius` of (qx, qy).
   * Returns an array of neighbor indices.
   * More convenient than queryState when you need a list, not a callback.
   */
  query(qx, qy, radius, state) {
    const results = [];
    this.queryState(qx, qy, radius, state, (j, _d2) => {
      results.push(j);
    });
    return results;
  }

  /**
   * Find all particles within `radius` of (qx, qy).
   * Calls callback(neighborIndex, distSquared) for each neighbor.
   * Distances are SQUARED — caller compares with radius*radius.
   */
  queryState(qx, qy, radius, state, callback) {
    const cs = this.cellSize;
    const cols = this.cols;
    const rows = this.rows;
    const r2 = radius * radius;

    const minCol = Math.max(0, ((qx - radius) / cs) | 0);
    const maxCol = Math.min(cols - 1, ((qx + radius) / cs) | 0);
    const minRow = Math.max(0, ((qy - radius) / cs) | 0);
    const maxRow = Math.min(rows - 1, ((qy + radius) / cs) | 0);

    for (let row = minRow; row <= maxRow; row++) {
      for (let col = minCol; col <= maxCol; col++) {
        let idx = this.cellStart[row * cols + col];
        while (idx !== -1) {
          const dx = qx - state.x[idx];
          const dy = qy - state.y[idx];
          const d2 = dx * dx + dy * dy;
          if (d2 < r2 && d2 > 0) {
            callback(idx, d2);
          }
          idx = this.particleNext[idx];
        }
      }
    }
  }
}

File: hard/src/simulation/spatial_hash.py
"""
GPU-native spatial hash with bucket capacity hard limit.

Uses cell_head + particle_next linked-list scheme with atomic operations.
Bucket capacity hard limit prevents O(N^2) collapse when particles clump.
"""

import taichi as ti


@ti.data_oriented
class SpatialHash:
    """GPU spatial hash grid with bucket capacity limit."""

    def __init__(self, cfg: dict):
        self.world_w = cfg['world']['width']
        self.world_h = cfg['world']['height']
        self.cell_size = cfg['world']['cell_size']
        self.bucket_max = cfg['simulation']['bucket_max']
        self.n_particles = cfg['simulation']['num_particles']

        self.cols = int(self.world_w / self.cell_size) + 1
        self.rows = int(self.world_h / self.cell_size) + 1
        self.num_cells = self.cols * self.rows

        # Linked-list structure
        self.cell_head = ti.field(dtype=ti.i32, shape=self.num_cells)
        self.particle_next = ti.field(dtype=ti.i32, shape=self.n_particles)
        self.cell_count = ti.field(dtype=ti.i32, shape=self.num_cells)

    @ti.kernel
    def build(self, pos_x: ti.template(), pos_y: ti.template()):
        """Build spatial hash from current particle positions.

        Accepts Taichi fields (GPU, zero-copy) directly — no CPU-GPU transfer.
        Previously accepted numpy arrays which forced a to_numpy() round-trip.
        """
        # Reset
        for cell in range(self.num_cells):
            self.cell_head[cell] = -1
            self.cell_count[cell] = 0

        # Insert particles with bucket capacity limit
        for i in range(self.n_particles):
            col = int(pos_x[i] / self.cell_size)
            row = int(pos_y[i] / self.cell_size)
            # Clamp to valid range
            col = ti.math.clamp(col, 0, self.cols - 1)
            row = ti.math.clamp(row, 0, self.rows - 1)
            cell = row * self.cols + col

            count = ti.atomic_add(self.cell_count[cell], 1)
            if count < self.bucket_max:
                # Insert at head of linked list
                old = self.cell_head[cell]
                self.cell_head[cell] = i
                self.particle_next[i] = old
            else:
                # Bucket full — particle excluded from neighbor queries this step
                self.particle_next[i] = -1

    @ti.func
    def query_neighbors(self, qx: ti.f32, qy: ti.f32, radius: ti.f32,
                        pos_x: ti.types.ndarray(), pos_y: ti.types.ndarray(),
                        neighbor_indices: ti.types.ndarray(),
                        neighbor_dists: ti.types.ndarray()) -> ti.i32:
        """
        Find all particles within radius of (qx, qy).
        Returns count. Results written to neighbor_indices and neighbor_dists.
        """
        r2 = radius * radius
        count = 0

        min_col = ti.math.clamp(int((qx - radius) / self.cell_size), 0, self.cols - 1)
        max_col = ti.math.clamp(int((qx + radius) / self.cell_size), 0, self.cols - 1)
        min_row = ti.math.clamp(int((qy - radius) / self.cell_size), 0, self.rows - 1)
        max_row = ti.math.clamp(int((qy + radius) / self.cell_size), 0, self.rows - 1)

        for row in range(min_row, max_row + 1):
            for col in range(min_col, max_col + 1):
                cell = row * self.cols + col
                idx = self.cell_head[cell]
                while idx != -1:
                    dx = qx - pos_x[idx]
                    dy = qy - pos_y[idx]
                    d2 = dx * dx + dy * dy
                    if d2 < r2 and d2 > 0.0:
                        if count < 128:  # max neighbors per query
                            neighbor_indices[count] = idx
                            neighbor_dists[count] = ti.sqrt(d2)
                            count += 1
                    idx = self.particle_next[idx]

        return count
